services/agent/tests/unit/test_volumes_cleanup_s3.py

```diff
@@ -0,0 +1,167 @@
+            key_path = f"{swarm_stack_name}/{study_id}/{node_uuid}/{run_id}/"
+            if s3_object.key.startswith(key_path):
+                file_object = await s3_object.get()
+                file_path: Path = save_to / s3_object.key.replace(key_path, "")
+                file_path.parent.mkdir(parents=True, exist_ok=True)
+                print(f"Saving file to {file_path}")
+                file_content = await file_object["Body"].read()
+                file_path.write_bytes(file_content)
+
+
+def _create_data(folder: Path) -> None:
+    for file in {  # pylint:disable=use-sequence-for-iteration
+        ".hidden_do_not_remove",
+        "key_values.json",
+        "f1.txt",
+        "f2.txt",
+        "f3.txt",
+        "d1/f1.txt",
+        "d1/f2.txt",
+        "d1/f3.txt",
+        "d1/sd1/f1.txt",
+        "d1/sd1/f2.txt",
+        "d1/sd1/f3.txt",
+    }:
+        file_path = folder / file
+        file_path.parent.mkdir(parents=True, exist_ok=True)
+        file_path.write_text("test")
+
+
+# FIXTURES
+
+
+@pytest.fixture
+def save_to(tmp_path: Path) -> Path:
+    return tmp_path / "save_to"
+
+
+# TESTS
+
+
+async def test_get_s3_path(
+    unused_volume: DockerVolume,
+    swarm_stack_name: str,
+    study_id: str,
+    node_uuid: str,
+    run_id: str,
+    bucket: str,
+):
+    volume_data = await unused_volume.show()
+    assert _get_s3_path(bucket, volume_data["Labels"]) == Path(
+        f"/{bucket}/{swarm_stack_name}/{study_id}/{node_uuid}/{run_id}"
+    )
+
+
+async def test_store_to_s3(
+    unused_volume: DockerVolume,
+    mocked_s3_server_url: HttpUrl,
+    unused_volume_path: Path,
+    save_to: Path,
+    study_id: str,
+    node_uuid: str,
+    run_id: str,
+    bucket: str,
+    settings: ApplicationSettings,
+):
+    _create_data(unused_volume_path)
+    dyv_volume = await unused_volume.show()
+
+    # overwrite to test locally not against volume
+    # root permissions are required to access this
+    dyv_volume["Mountpoint"] = unused_volume_path
+
+    await store_to_s3(
+        dyv_volume=dyv_volume,
+        s3_access_key="xxx",
+        s3_secret_key="xxx",
+        s3_bucket=bucket,
+        s3_endpoint=mocked_s3_server_url,
+        s3_region="us-east-1",
+        s3_provider=S3Provider.MINIO,
+        s3_parallelism=3,
+        s3_retries=1,
+        exclude_files=settings.AGENT_VOLUMES_CLEANUP_EXCLUDE_FILES,
+    )
+
+    await _download_files_from_bucket(
+        endpoint=mocked_s3_server_url,
+        access_key="xxx",
+        secret_key="xxx",
+        bucket_name=bucket,
+        save_to=save_to,
+        swarm_stack_name=dyv_volume["Labels"]["swarm_stack_name"],
+        study_id=study_id,
+        node_uuid=node_uuid,
+        run_id=run_id,
+    )
+
+    hashes_on_disk = _get_file_hashes_in_path(
+        unused_volume_path, set(map(Path, settings.AGENT_VOLUMES_CLEANUP_EXCLUDE_FILES))
+    )
+    hashes_in_s3 = _get_file_hashes_in_path(save_to)
```

Comment:
question... who is cleaning up all these files?
Can you maybe use tmp_path fixture somewhere? I just spent the other day 3 hours to recover my disk because of some test